lsp/src/Language/LSP/Server/Processing.hs

```diff
@@ -449,31 +451,30 @@ handle' ::
   TClientMessage meth ->
   m ()
 handle' logger mAction m msg = do
-  maybe (return ()) (\f -> f msg) mAction
+  shutdown <- isShuttingDown
+  let allowedMethod m = case (splitClientMethod m, m) of
+        (IsClientNot, SMethod_Exit) -> True
+        (IsClientReq, SMethod_Shutdown) -> True
+        _ -> False
+
+  when (not shutdown || allowedMethod m) $ maybe (return ()) (\f -> f msg) mAction
```

Comment:
I think we don't need to add to the note to explain this, since we already say we ignore notifications, so it's reasonable to infer that we don't update our state or anything.